File: go/lasagna-master/lasagna_master.go
package lasagna

// PreparationTime returns how long it takes to prepare.
func PreparationTime(layers []string, time int) int {
	if time == 0 {
		time = 2
	}
	return len(layers) * time
}

// Quantities returns the quantity of noodles and sauce needed.
func Quantities(layers []string) (int, float64) {
	var noodles int
	var sauce float64
	for _, contents := range layers {
		if contents == "noodles" {
			noodles += 50
		} else if contents == "sauce" {
			sauce += 0.2
		}
	}
	return noodles, sauce
}

// AddSecretIngredient adds an ingredient to the list.
func AddSecretIngredient(friendList, myList []string) {
	myList[len(myList)-1] = friendList[len(friendList)-1]
}

// ScaleRecipe scales recipe amounts.
func ScaleRecipe(amounts []float64, portions int) []float64 {
	out := make([]float64, 0, len(amounts))
	scale := float64(portions) / 2
	for _, amount := range amounts {
		out = append(out, amount*scale)
	}
	return out
}

File: javascript/lasagna-master/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check

/**
 * Implement the functions needed to solve the exercise here.
 * Do not forget to export them so they are available for the
 * tests. Here an example of the syntax as reminder:
 *
 * export function yourFunction(...) {
 *   ...
 * }
 */

export function cookingStatus(elapsed) {
  switch (elapsed) {
    case undefined:
      return 'You forgot to set the timer.'
    case 0:
      return 'Lasagna is done.'
    default:
      return 'Not done, please wait.'
  }
}

export function preparationTime(layers, perLayer = 2) {
  return layers.length * perLayer
}

export function quantities(layers) {
  let layerCount = {}
  for (const layer of layers) {
    layerCount[layer] = (layerCount[layer] ?? 0) + 1
  }
  console.log(layerCount)
  return {noodles: (layerCount.noodles ?? 0) * 50, sauce: (layerCount.sauce ?? 0) * 0.2}
}

export function addSecretIngredient(friendsList, myList) {
  myList.push(friendsList.slice(-1)[0])
}

export function scaleRecipe(recipe, size) {
  let scaled = {}
  for (const i in recipe) {
    scaled[i] = recipe[i] * size / 2
  }
  return scaled
}
